mesos-client/src/main/scala/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -0,0 +1,514 @@
+    .map { case (ctx, call) =>
+      (ctx, call.update(
+        _.optionalFrameworkId := Some(ctx.frameworkId.getOrElse(throw new IllegalStateException("FrameworkID not set"))))
+      )
+    }
+
+  private val sinkHub: RunnableGraph[Sink[Call, NotUsed]] =
+    MergeHub.source[Call](perProducerBufferSize = 16)
+      .mapAsync(1)(event => contextPromise.future.map(ctx => (ctx, event)))
+      .via(callEnhancer)
+      .via(log("Sending "))
+      .via(eventSerializer)
+      .via(requestBuilder)
+      .to(sink)
+
+  // By running/materializing the consumer we get back a Sink, and hence now have access to feed elements into it.
+  // This Sink can be materialized any number of times, and every element that enters the Sink will be consumed by
+  // our consumer.
+  override val mesosSink: Sink[Call, NotUsed] = sinkHub.run()
+}
+
+trait MesosApi {
+
+  /**
+    * First call to this method will initialize the connection to mesos and return a `Source[String, NotUser]` with
+    * mesos `Event`s. All subsequent calls will return the previously created event source.
+    * The connection is initialized with a `POST /api/v1/scheduler` with the framework info in the body. The request
+    * is answered by a `SUBSCRIBED` event which contains `MesosStreamId` header. This is reused by all later calls to
+    * `/api/v1/scheduler`.
+    * Multiple subscribers can attach to returned `Source[String, NotUsed]` to receive mesos events. The stream will
+    * be closed either on connection error or connection shutdown e.g.:
+    * ```
+    * client.source.runWith(Sink.ignore).onComplete{
+    *  case Success(res) => logger.info(s"Stream completed: $res")
+    *  case Failure(e) => logger.error(s"Error in stream: $e")
+    * }
+    * ```
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#subscribe-1
+    */
+  def mesosSource: Source[Event, NotUsed]
+
+  /**
+    * Sink for mesos calls. Multiple publishers can materialize this sink to send mesos `Call`s. Every `Call` is sent
+    * using a new HTTP connection.
+    * Note: a scheduler can't send calls to mesos without subscribing first (see [MesosClient.source] method). Calls
+    * published to sink without a successful subscription will be buffered and will have to wait for subscription
+    * connection. Always call `source()` first.
+    *
+    * http://mesos.apache.org/documentation/latest/scheduler-http-api/#calls
+    */
+  def mesosSink: Sink[Call, NotUsed]
+
+  /**
+    * A kill switch for the mesos source. Calling `shutdown()` or `abort()` on it will close the connection to mesos.
+    * Note that depending on `failoverTimeout` provided with SUBSCRIBED call mesos could start killing tasks and
+    * executors started by the framework. Make sure to set `failoverTimeout` appropriately. See `teardown()` method
+    * for another way to shutdown a framework.
+    */
+  val killSwitch: UniqueKillSwitch
```

Comment:
What is the advantage of exposing two ways to tear down?

Reply:
I'm exposing the `killSwitch` here and it offers 2 ways of tearing down (`shutdown` completes the downstream, `abort` fails it).

Reply:
Are you aware of mapMaterializedValue? It seems like we could expose the killSwitch as a materialized value (whereas right now the materialized value for MesosSource is NotUsed)